lib/src/client/http2_connection.dart

```diff
@@ -43,7 +43,6 @@ class Http2ClientConnection implements connection.ClientConnection {
 
   connection.ConnectionState _state = ConnectionState.idle;
 
-  @visibleForTesting
   void Function(Http2ClientConnection connection)? onStateChanged;
```

Comment:
I feel like you could still add `onStateChanged` to `ClientConnection`:

```dart
class ClientConnection {
  /// Set state change listener for this connection. The given callback will be 
  /// invoked when the state of this connection changes.
  set onStateChanged(void Function(ConnectionState) cb);
}
```

(Note: I changed signature because this will make code simpler.).

This setter can be (for now) an no-op for XHR based connection. 

This would allow you to avoid any platform specific and duplicated code in `lib/src/client/http2_channel.dart`.